### PR TITLE
Add support for additional DWT counters

### PR DESCRIPTION
**Pull request by newAM**
This adds support for these DWT counters:
* CPI counter
* Exception overhead counter
* LSU counter
* Folded-instruction counter

The getters do not have the `get_` prefix per the [C-GETTER](https://rust-lang.github.io/api-guidelines/naming.html#c-getter) API guidelines. This does not match the existing `get_cycle_count` method. I was not sure if I should maintain consistency or not, let me know if you want this changed.

Here is the documentation to validate each of the added methods:
* [ARMv7-M DWT register summary](https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/DWT-register-summary)
    * [Control Register](https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/Control-register--DWT-CTRL)
        * `num_comp`
            * [ARMv6-M DWT register summary](https://developer.arm.com/documentation/ddi0419/c/Debug-Architecture/ARMv6-M-Debug/The-Data-Watchpoint-and-Trace-unit/DWT-register-summary)
            * The `num_comp` method is not guarded with `#[cfg(not(armv6m))]` because it is valid for both.
        * `has_exception_trace`
        * `has_external_match`
        * `has_cycle_counter`
        * `has_profiling_counter`
        * `disable_cycle_counter`
        * `cycle_counter_enabled`
    * [CPI count register](https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/CPI-Count-register--DWT-CPICNT)
        * `cpi_count`
        * `set_cpi_count`
    * [Exception Overhead Count register](https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/Exception-Overhead-Count-register--DWT-EXCCNT)
        * `exception_count`
        * `set_exception_count`
    * [Sleep Count register](https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/Sleep-Count-register--DWT-SLEEPCNT)
        * `sleep_count`
        * `set_sleep_count`
    * [LSU Count register](https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/LSU-Count-register--DWT-LSUCNT)
        * `lsu_count`
        * `set_lsu_count`
    * [Folded-instruction Count register](https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/Folded-instruction-Count-register--DWT-FOLDCNT)
        * `fold_count`
        * `set_fold_count`

### CODE REVIEW

File: src/peripheral/dwt.rs
Comment (thejpster) [inline src/peripheral/dwt.rs:65]: It might be worth trying to add a link to the Arm Armv7-EM ARM here, to explain where the constants came from.

https://developer.arm.com/documentation/ddi0403/d/Debug-Architecture/ARMv7-M-Debug/The-Data-Watchpoint-and-Trace-unit/Control-register--DWT-CTRL might work.

**Comment by thejpster** [comment]
I think I'd like to see a `cycle_count()` method for consistency. I'm relaxed as to whether 'get_cycle_count()` is marked deprecated, or on what timescale (but it should at least note in the docs that `cycle_count()` is canonical).

**Comment by newAM** [comment]
> I think I'd like to see a cycle_count() method for consistency. I'm relaxed as to whether get_cycle_count() is marked deprecated, or on what timescale (but it should at least note in the docs that cycle_count() is canonical).

I am happy to do that too.  This is a debug feature, the crate is still pre-1.0, and it is a minor change, so I don't think the impact will be significant.  Is it ok to make that change within this pull-request?  That is marking `get_cycle_count` as deprecated and adding a `cycle_count` method that does the same thing.  Not sure if there is any process for making a breaking change with this crate.

**Comment by thejpster** [comment]
I'll defer to someone like @therealprof or @adamgreig on that one.

**Comment by adamgreig** [comment]
Our policy is to make a minor release for breaking changes while <1.0 (per standard Rust semver), so I'm happy to deprecate `get_cycle_count` now, release that in the next cortex-m 0.7.x release, and remove it in the eventual 0.8.x (or 1.x) release.

**Review by adamgreig — COMMENTED** [review]
Thanks! This all LGTM, but could you update the changelog to reflect the deprecation of `get_cycle_count()`? Happy to merge after that.

**Comment by newAM** [comment]
> Thanks! This all LGTM, but could you update the changelog to reflect the deprecation of `get_cycle_count()`? Happy to merge after that.

Whoops, good catch, thanks!  I updated the changelog.

**Review by adamgreig — APPROVED** [review]
Thanks!

bors merge

**Pull request merged**
